### PR TITLE
sys/ztimer: allow compensation of ztimer_sleep() overhead

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

```ztimer_sleep()``` can take considerable overhead.

This PR introduces a per-clock field ```adjust_sleep``` that allows compensating this overhead.
It can be configured for ztimer_usec by setting CONFIG_ZTIMER_USEC_ADJUST_SLEEP.

For consistency reasons, the corresponding adjust value field for "ztimer_set()" has been renamed from just "adjust" to "adjust_set", the config option CONFIG_ZTIMER_USEC_ADJUST_SET.

The sleep value is used *in addition to the set value*.

### Testing procedure

- run tests/ztimer_overhead, not down "min" value, recompile with ```CFLAGS += -DCONFIG_ZTIMER_USEC_ADJUST_SET=<min>```

- run tests/xtimer_usleep compiled with ```CFLAGS += -DCONFIG_ZTIMER_USEC_ADJUST_SET=<min>``` and ```USEMODULE+=ztimer_xtimer_compat```, note offset value

- re-run tests/xtimer_usleep compiled with ```CFLAGS += -DCONFIG_ZTIMER_USEC_ADJUST_SET=<min>``` and ```USEMODULE+=ztimer_xtimer_compat```, and ```CFLAGS += -DCONFIG_ZTIMER_USEC_ADJUST_SLEEP=<offset>```

The offset should now be closer to zero.

### CODE REVIEW

**Comment by kaspar030** [comment]
btw, this is a first step towards making this automatic in ztimer's auto_init().

File: tests/ztimer_overhead/main.c [hunk 1]
```diff
@@ -37,8 +37,8 @@ int main(void)
     int32_t max = INT32_MIN;
 
     /* unset configured adjustment */
-    /* ZTIMER_USEC->adjust = 0; */
-    printf("ZTIMER_USEC->adjust = %" PRIu32 "\n", ZTIMER_USEC->adjust);
+    /* ZTIMER_USEC->adjust_set = 0; */
+    printf("ZTIMER_USEC->adjust_set = %" PRIu32 "\n", ZTIMER_USEC->adjust_set);
```
Comment (fjmolinas) [inline tests/ztimer_overhead/main.c:41]:
```suggestion
    printf("ZTIMER_USEC->adjust_set = %" PRIu16 "\n", ZTIMER_USEC->adjust_set);
```

**Comment by fjmolinas** [comment]
Some testing:


- `BOARD=iotlab-m3 make -C tests/ztimer_overhead/ flash test`

```
main(): This is RIOT! (Version: 2020.10-devel-1222-g2ed879-pr-14936)
ZTIMER_USEC->adjust_set = 0
min=9 max=9 avg_diff=9
```

- `CFLAGS=-DCONFIG_ZTIMER_USEC_ADJUST_SET=9 BOARD=iotlab-m3 make -C tests/ztimer_overhead/ flash test`

```
main(): This is RIOT! (Version: 2020.10-devel-1222-g2ed879-pr-14936)
ZTIMER_USEC->adjust_set = 9
min=0 max=0 avg_diff=0
```

- `USEMODULE=ztimer_xtimer_compat CFLAGS=-DCONFIG_ZTIMER_USEC_ADJUST_SET=9 BOARD=iotlab-m3 make -C tests/xtimer_usleep/ flash test`

```
main(): This is RIOT! (Version: 2020.10-devel-1222-g2ed879-pr-14936)
Running test 5 times with 7 distinct sleep times
Slept for 10010 us (expected: 10000 us) Offset: 10 us
Slept for 50010 us (expected: 50000 us) Offset: 10 us
Slept for 10244 us (expected: 10234 us) Offset: 10 us
Slept for 56790 us (expected: 56780 us) Offset: 10 us
Slept for 12132 us (expected: 12122 us) Offset: 10 us
Slept for 98775 us (expected: 98765 us) Offset: 10 us
Slept for 75010 us (expected: 75000 us) Offset: 10 us
Slept for 10010 us (expected: 10000 us) Offset: 10 us
Slept for 50010 us (expected: 50000 us) Offset: 10 us
Slept for 10244 us (expected: 10234 us) Offset: 10 us
Slept for 56790 us (expected: 56780 us) Offset: 10 us
Slept for 12132 us (expected: 12122 us) Offset: 10 us
Slept for 98775 us (expected: 98765 us) Offset: 10 us
Slept for 75010 us (expected: 75000 us) Offset: 10 us
Slept for 10010 us (expected: 10000 us) Offset: 10 us
Slept for 50010 us (expected: 50000 us) Offset: 10 us
Slept for 10244 us (expected: 10234 us) Offset: 10 us
Slept for 56790 us (expected: 56780 us) Offset: 10 us
Slept for 12132 us (expected: 12122 us) Offset: 10 us
Slept for 98775 us (expected: 98765 us) Offset: 10 us
Slept for 75010 us (expected: 75000 us) Offset: 10 us
Slept for 10010 us (expected: 10000 us) Offset: 10 us
Slept for 50010 us (expected: 50000 us) Offset: 10 us
Slept for 10244 us (expected: 10234 us) Offset: 10 us
Slept for 56790 us (expected: 56780 us) Offset: 10 us
Slept for 12132 us (expected: 12122 us) Offset: 10 us
Slept for 98775 us (expected: 98765 us) Offset: 10 us
Slept for 75010 us (expected: 75000 us) Offset: 10 us
Slept for 10010 us (expected: 10000 us) Offset: 10 us
Slept for 50010 us (expected: 50000 us) Offset: 10 us
Slept for 10244 us (expected: 10234 us) Offset: 10 us
Slept for 56790 us (expected: 56780 us) Offset: 10 us
Slept for 12132 us (expected: 12122 us) Offset: 10 us
Slept for 98775 us (expected: 98765 us) Offset: 10 us
Slept for 75010 us (expected: 75000 us) Offset: 10 us
Test ran for 1606222 us
```

- `USEMODULE=ztimer_xtimer_compat CFLAGS="-DCONFIG_ZTIMER_USEC_ADJUST_SET=9 -DCONFIG_ZTIMER_USEC_ADJUST_SLEEP=10" BOARD=iotlab-m3 make -C tests/xtimer_usleep/ flash test`

```
Slept for 10000 us (expected: 10000 us) Offset: 0 us
Slept for 50000 us (expected: 50000 us) Offset: 0 us
Slept for 10234 us (expected: 10234 us) Offset: 0 us
Slept for 56780 us (expected: 56780 us) Offset: 0 us
Slept for 12122 us (expected: 12122 us) Offset: 0 us
Slept for 98765 us (expected: 98765 us) Offset: 0 us
Slept for 75000 us (expected: 75000 us) Offset: 0 us
Slept for 10000 us (expected: 10000 us) Offset: 0 us
Slept for 50000 us (expected: 50000 us) Offset: 0 us
Slept for 10234 us (expected: 10234 us) Offset: 0 us
Slept for 56780 us (expected: 56780 us) Offset: 0 us
Slept for 12122 us (expected: 12122 us) Offset: 0 us
Slept for 98765 us (expected: 98765 us) Offset: 0 us
Slept for 75000 us (expected: 75000 us) Offset: 0 us
Slept for 10000 us (expected: 10000 us) Offset: 0 us
Slept for 50000 us (expected: 50000 us) Offset: 0 us
Slept for 10234 us (expected: 10234 us) Offset: 0 us
Slept for 56780 us (expected: 56780 us) Offset: 0 us
Slept for 12122 us (expected: 12122 us) Offset: 0 us
Slept for 98765 us (expected: 98765 us) Offset: 0 us
Slept for 75000 us (expected: 75000 us) Offset: 0 us
Slept for 10000 us (expected: 10000 us) Offset: 0 us
Slept for 50000 us (expected: 50000 us) Offset: 0 us
Slept for 10234 us (expected: 10234 us) Offset: 0 us
Slept for 56780 us (expected: 56780 us) Offset: 0 us
Slept for 12122 us (expected: 12122 us) Offset: 0 us
Slept for 98765 us (expected: 98765 us) Offset: 0 us
Slept for 75000 us (expected: 75000 us) Offset: 0 us
Slept for 10000 us (expected: 10000 us) Offset: 0 us
Slept for 50000 us (expected: 50000 us) Offset: 0 us
Slept for 10234 us (expected: 10234 us) Offset: 0 us
Slept for 56780 us (expected: 56780 us) Offset: 0 us
Slept for 12122 us (expected: 12122 us) Offset: 0 us
Slept for 98765 us (expected: 98765 us) Offset: 0 us
Slept for 75000 us (expected: 75000 us) Offset: 0 us
Test ran for 1604972 us
```

In the test I had to change :

```diff
diff --git a/tests/xtimer_usleep/tests/01-run.py b/tests/xtimer_usleep/tests/01-run.py
index f629096073..8bd57295c8 100755
--- a/tests/xtimer_usleep/tests/01-run.py
+++ b/tests/xtimer_usleep/tests/01-run.py
@@ -35,7 +35,7 @@ def testfunc(child):
                 sleep_time = int(child.match.group(1))
                 exp = int(child.match.group(2))
                 upper_bound = exp + (exp * INTERNAL_JITTER)
-                if not (exp < sleep_time < upper_bound):
+                if not (exp <= sleep_time < upper_bound):
                     delta = (upper_bound-exp)
                     error = min(upper_bound-sleep_time, sleep_time-exp)
                     raise InvalidTimeout("Invalid timeout %d, expected %d < timeout < %d"
```

**Comment by fjmolinas** [comment]
@kaspar030 some nitpicks otherwise it seems to work as advertised, one comment though regarding the `CONFIG_ZTIMER_USEC_ADJUST_SLEEP`, is there a risk of it undersleeping? Should the value be set to the offset found with the test `-1`?

**Comment by kaspar030** [comment]
> one comment though regarding the `CONFIG_ZTIMER_USEC_ADJUST_SLEEP`, is there a risk of it undersleeping? Should the value be set to the offset found with the test `-1`?

yes there is. it should be set so "min" is 0, to guarantee "at least X" semantics. well, I think we specified this to be +- one, so -1 in some rare cases might still be acceptable. not sure.

**Comment by kaspar030** [comment]
> In the test I had to change :

fixed. interesting, the test previously failed on zero offset...

File: tests/ztimer_overhead/main.c (hunk 1 above)
Comment (kaspar030) [reply tests/ztimer_overhead/main.c]: fixed

**Comment by fjmolinas** [comment]
> > one comment though regarding the `CONFIG_ZTIMER_USEC_ADJUST_SLEEP`, is there a risk of it undersleeping? Should the value be set to the offset found with the test `-1`?
> 
> yes there is. it should be set so "min" is 0, to guarantee "at least X" semantics. well, I think we specified this to be +- one, so -1 in some rare cases might still be acceptable. not sure.

Would it be possible to add a short mention to the `ztimer.h` on how these `adjust` values can be found out and how to be conservative when selecting them (this +-1 policy for example.)

**Comment by kaspar030** [comment]
> Would it be possible to add a short mention to the `ztimer.h` on how these `adjust` values can be found out and how to be conservative when selecting them (this +-1 policy for example.)

I intend to do more docs on this in the follow-up PR, which will hopefully just figure out the values automagically on each boot.

**Comment by fjmolinas** [comment]
> I intend to do more docs on this in the follow-up PR, which will hopefully just figure out the values automagically on each boot.

OK, but I intend to quote you on this!

**Comment by fjmolinas** [comment]
@kaspar030 please squash!

**Comment by kaspar030** [comment]
all green!

**Review by fjmolinas — APPROVED** [review]
ACK.

**Comment by kaspar030** [comment]
thanks for the review!

**Pull request merged**